src/solvers/dg_tree/dg.jl

```diff
@@ -268,19 +333,25 @@ include("basis_lobatto_legendre.jl")
           surface_flux=flux_central,
           volume_integral::AbstractVolumeIntegral=VolumeIntegralWeakForm(),
           mortar=MortarL2(basis))
+    DGSEM(; RealT=Float64, polydeg::Integer,
+            surface_flux=flux_central,
+            volume_integral::AbstractVolumeIntegral=VolumeIntegralWeakForm(),
+            mortar=MortarL2(basis))
 
 Create a discontinuous Galerkin spectral element method (DGSEM) using a
 [`LobattoLegendreBasis`](@ref) with polynomials of degree `polydeg`.
 """
-const DGSEM = DG{Basis, Mortar, SurfaceFlux, VolumeIntegral} where {Basis<:LobattoLegendreBasis, Mortar, SurfaceFlux, VolumeIntegral}
+const DGSEM = DG{Basis, Mortar, SurfaceIntegralWeakForm{SurfaceFlux}, VolumeIntegral} where {Basis<:LobattoLegendreBasis, Mortar, SurfaceFlux, VolumeIntegral}
+# const DGSEM = DG{Basis, Mortar, SurfaceIntegral, VolumeIntegral} where {Basis<:LobattoLegendreBasis, Mortar, SurfaceIntegral<:SurfaceIntegralWeakForm, VolumeIntegral}
 
 function DGSEM(basis::LobattoLegendreBasis,
                surface_flux=flux_central,
                volume_integral::AbstractVolumeIntegral=VolumeIntegralWeakForm(),
                mortar=MortarL2(basis))
 
-  return DG{typeof(basis), typeof(mortar), typeof(surface_flux), typeof(volume_integral)}(
-    basis, mortar, surface_flux, volume_integral)
+  surface_integral = SurfaceIntegralWeakForm(surface_flux)
```

Comment:
OK, so DGSEM is always considered to be using `SurfaceIntegralWeakForm`? I think then some of my comments above do not make so much sense anymore :-/ I think I do not yet 100% understand the difference between strong form and weak form surface integral, maybe some clarification (possibly via different communication channels) is required

Reply:
That part will become more clear once the FD methods come to Trixi. I do not want to document it right now (without having an example code running) - that's why this part is less specified. The basic difference is that the weak form volume integral contains only the numerical flux while the strong form volume integral contains the difference of the numerical flux and the `flux`. The reason for this is integration by parts, which is used only for the weak form volume integral/terms. We still use the weak form surface integral/terms for the other volume integrals/terms, since we subtract surface contributions explicitly from the derivative matrix. That's why we use `derivative_split` instead of `derivative_matrix` everywhere.

Reply:
A strong form has basically 3 parts that need to be computed: 

RHS = volume  + surface1 + surface2

volume: this only depends on the element local data. It computes the (discrete) L2 projection of the flux divergence onto the basis. If we assume collocation, then the term basically reduces to the application of the D matrix. The implementation and structure of this term is identical to the weak form volume, except that the D matrix is changed to the appropriate operator of the weak form.

surface1: this term stems from integration-by-parts once, it is an integral over the surface of the flux normal component. This term uses a numerical flux to connect to the neighbor elements. Numerical fluxes typically depend on data from the two elements connected via the surface. Hence, it depends on neighbor data, hence it needs communication/data connectivity before it can be computed. This term can be computed once and applied (with different signs) to both elements, connected via the corresponding surface. This term is identical to the surface term in the weak form. 

surface2: this is a term that stems from integration-by-parst twice, it is an integral over the surface of the flux normal component - but only its "inner" component. "Inner" means, that it is not uniquely defined at the surface, it depends on the considered element and thus needs to be computed for either side/element connected through the surface. The good thing is, that it is thus element local - it only needs element data and hence no communication/connectivity. If we assume collocation and assume that the collocation nodes include the nodes at the element interfaces, the term can be directly computed with the B matrix (of the SBP property).

There are two choices to implement surface2 (maybe more different ways?): 

(i) The calculation of surface2 can thus be either done within the calculation of surface1 (one needs to be careful to compute two different surface contributions for each element connected via the surface). In this case, the routine needs to know which input value left and right belongs to which element connected to the surface. Note: having to manage two different surface contributions (one for the left element, one for the right element) is a good strategy for non-conservative terms (such as in MHD) as well. 

(ii) Or within the calculation of volume. A naive way would be to just do the surface integral routine again, but only with local fluxes. This is inefficient.  Often, it is possible to modify the D matrix in the strong form volume integral to directly account for this by assuming collocation of the fluxes and that the nodes used to compute the volume integral include also surface nodes (This is what we have done for the split-form volume integral: we just modified the D matrix to 2*D-B to include surface2). The advantage with this implementation is, that there is a clear difference between terms that need communication and terms that are local. 

Last comment: if the operator has the SBP property, of course, weak form and strong form is identical (that is basically the definition of SBP). Hence, translated to the terms used in this post here, for all our schemes we have: 

volume(weak form) = volume(strong form) + surface2

Reply:
Great description! My reason for introducing the strong form here is that we take SBP FD operators from SummationByPartsOperators.jl, which implement the strong for volume terms/integral efficiently but do not provide the modified strong form `derivative_split` we use for DGSEM in Trixi. The current implementation of the strong form surface terms/integral follows your strategy i).

Reply:
I agree, this was a very nice breakdown of the different strategies / naming conventions of the DG flavours.

Reply:
I added it as a TODO note to the list in #607

Reply:
Thanks!